src/common/cockpitauthorize.c

```diff
@@ -0,0 +1,577 @@
+}
+
+const char *
+cockpit_authorize_subject (const char *challenge,
+                           char **subject)
+{
+  size_t len;
+
+  challenge = cockpit_authorize_type (challenge, NULL);
+  if (!challenge)
+    return NULL;
+
+  len = strcspn (challenge, ": ");
+  if (len == 0)
+    {
+      message ("invalid \"authorize\" message \"challenge\": no subject");
+      errno = EINVAL;
+      return NULL;
+    }
+
+  if (subject)
+    {
+      *subject = strndup (challenge, len);
+      if (!*subject)
+        {
+          message ("couldn't allocate memory for \"authorize\" message \"challenge\"");
+          errno = ENOMEM;
+          return NULL;
+        }
+    }
+
+  if (challenge[len])
+    len++;
+  while (challenge[len] == ' ')
+    len++;
+  return challenge + len;
+}
+
+char *
+cockpit_authorize_parse_basic (const char *challenge,
+                               char **user)
+{
+  unsigned char *buf = NULL;
+  char *type = NULL;
+  size_t len;
+  ssize_t res;
+  int errn = 0;
+  size_t off;
+
+  challenge = cockpit_authorize_type (challenge, &type);
+  if (!challenge)
+    return NULL;
+
+  if (strcmp (type, "basic") != 0)
+    {
+      message ("invalid prefix in Basic header");
+      errn = EINVAL;
+      goto out;
+    }
+
+  len = strcspn (challenge, " ");
+  buf = malloc (len + 1);
+  if (!buf)
+    {
+      message ("couldn't allocate memory for Basic header");
+      errn = ENOMEM;
+      goto out;
+    }
+
+  /* No value */
+  if (len == 0)
+    {
+      buf[0] = 0;
+      if (user)
+        *user = NULL;
+      goto out;
+    }
+
+  /* Decode and find split point */
+  res = cockpit_base64_pton (challenge, len, buf, len);
+  if (res < 0)
+    {
+      message ("invalid base64 data in Basic header");
+      errn = EINVAL;
+      goto out;
+    }
+  assert (res <= len);
+  buf[res] = 0;
+
+  off = strcspn ((char *)buf, ":");
+  if (off == 0 || off == res)
+    {
+      message ("invalid base64 data in Basic header");
+      errn = EINVAL;
+      goto out;
+    }
+
+  if (user)
+    {
+      *user = strndup ((char *)buf, off);
+      if (!*user)
+        {
+          message ("couldn't allocate memory for user name");
+          errn = ENOMEM;
+          goto out;
+        }
+    }
+
+  memmove (buf, buf + off + 1, res - off);
+
+out:
+  free (type);
+  if (errn != 0)
+    {
+      errno = errn;
+      free (buf);
+      buf = NULL;
+    }
+  return (char *)buf;
+}
+
+char *
+cockpit_authorize_build_basic (const char *user,
+                               const char *password)
+{
+  char *content = NULL;
+  char *encoded = NULL;
+  char *response = NULL;
+  size_t elen, clen;
+  int errn = 0;
+
+  if (!user)
+    user = "";
+  if (!password)
+    password = "";
+
+  if (asprintf (&content, "%s:%s", user, password) < 0)
+    {
+      errn = errno;
+      message ("could not build basic response");
+      goto out;
+    }
+
+  clen = strlen (content);
+  elen = cockpit_base64_size (clen);
+  encoded = malloc (elen + 1);
```

Comment:
Do we need a malloc check here? Or maybe cockpit_base64_ntop should handle null targets?